8361103: java_lang_Thread::async_get_stack_trace does not properly protect JavaThread

alexmenkov opened the pull request:
The fix updates `java_lang_Thread::async_get_stack_trace()` (used by `java.lang.Thread.getStackTrace()` to get stack trace for platform and mounted virtual threads) to correctly use `ThreadListHandle` for thread protection.

Testing: tier1..5

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8361103](https://bugs.openjdk.org/browse/JDK-8361103): java_lang_Thread::async_get_stack_trace does not properly protect JavaThread (**Bug** - P3)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [811db312](https://git.openjdk.org/jdk/pull/26119/files/811db312b1d20987abe90f168eee06455bdce7e6)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26119/head:pull/26119` \
`$ git checkout pull/26119`

Update a local copy of the PR: \
`$ git checkout pull/26119` \
`$ git pull https://git.openjdk.org/jdk.git pull/26119/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26119`

View PR using the GUI difftool: \
`$ git pr show -t 26119`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26119.diff">https://git.openjdk.org/jdk/pull/26119.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26119#issuecomment-3033453532)
</details>

dholmes-ora reviewed (commented):
Generally looks good. A couple of minor nits/queries.

Thanks

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1882,20 +1882,21 @@ oop java_lang_Thread::park_blocker(oop java_thread) {
-    thread = java_lang_Thread::thread(carrier_thread);
-  } else {
-    thread = java_lang_Thread::thread(java_thread);
+    java_thread = java_lang_Thread::thread(carrier_thread);
```
Actually I think we need a further check here. If we get the carrier thread directly, we have not checked that it is actually protected by the TLH - that is normally done by `cv_internal_thread_to_JavaThread` but that doesn't know about virtual threads and carriers!). I need to check if we have to fix `cv_internal_thread_to_JavaThread` for the virtual thread case.

alexmenkov replied:
Well, we don't have TLH protection for carrier (and looks like JVMTI also doesn't care about carrier protection).
I think it may be useful to add virtual thread support to `ThreadsListHandle` (and update comments in threadSMR.hpp).
What about:
```
  bool cv_internal_thread_to_JavaThread(jobject jthread, JavaThread ** jt_pp, oop * thread_oop_p);

+  bool cv_oop_to_JavaThread(oop thread_oop, JavaThread** jt_pp);
+  bool cv_thread_or_carrier_to_JavaThread(oop thread_oop, JavaThread** jt_pp, bool* is_virtual_p = nullptr);
```
Also we need a way to check if vthread is mounted to JavaThread (to be checked in handshake), I think it should go to javaThread.hpp/.cpp:

```
  inline bool is_vthread_mounted() const;
+  inline bool is_vthread_mounted(oop vthread) const;
```
What do you think?
Would be nice to hear @dcubed-ojdk opinion (AFAIK he implemented threadSMR stuff)

alexmenkov replied:
> I was thinking simply(?) to change `cv_internal_thread_to_JavaThread` so that it extracts the carrier for a vthread and checks it is within the given threads-list.

I prefer to keep the current behavior for existing callers to be sure they are not affected.
Added extra argument to `cv_internal_thread_to_JavaThread` to return carrrier's JavaThread.

> > Also we need a way to check if specific vthread is mounted to JavaThread (to be checked in handshake)
> 
> Not sure that would be of general use as the answer could change the instant you return from the operation. Explicit checks for being mounted are done within the handshake/safepoint op code as needed.

Ok, kept it as it is

dholmes-ora commented:
I'm also unclear if this code in the handshake is sufficient for dealing with the case where we have a mounted virtual thread initially, but it is unmounted before the handshake gets to execute:
```
    if (java_lang_VirtualThread::is_instance(_java_thread())) {
        // if (thread->vthread() != _java_thread()) // We might be inside a System.executeOnCarrierThread
        const ContinuationEntry* ce = thread->vthread_continuation();
        if (ce == nullptr || ce->cont_oop(thread) != java_lang_VirtualThread::continuation(_java_thread())) {
          return; // not mounted
        }
      }
```
The commented line seems an inaccurate characterization as we could be executing a completely different virtual thread on this carrier now - which is not related to `executeOnCarrierThread`.

dholmes-ora reviewed (changes requested):
Revoking my approval as there may be further issues.

alexmenkov commented:
> I'm also unclear if this code in the handshake is sufficient for dealing with the case where we have a mounted virtual thread initially, but it is unmounted before the handshake gets to execute:
> 
> ```
>     if (java_lang_VirtualThread::is_instance(_java_thread())) {
>         // if (thread->vthread() != _java_thread()) // We might be inside a System.executeOnCarrierThread
>         const ContinuationEntry* ce = thread->vthread_continuation();
>         if (ce == nullptr || ce->cont_oop(thread) != java_lang_VirtualThread::continuation(_java_thread())) {
>           return; // not mounted
>         }
>       }
> ```
> 
> The commented line seems an inaccurate characterization as we could be executing a completely different virtual thread on this carrier now - which is not related to `executeOnCarrierThread`.

I suppose the comment is obsolete. I don't see `executeOnCarrierThread` in the current codebase.
The condition checks if there were some changes in vthread/carrier:
`ce == nullptr` - the carries has no mounted vthread now (i.e. the vthread has been unmounted);
`ce->cont_oop(thread) != java_lang_VirtualThread::continuation(_java_thread())` - the carrier has other vthread mounted;
Even if there were unmount/mount the same vthread on the carrier, it's fine (we need vthread stack trace and we are in handshake with its carrier)

alexmenkov commented:
- added mounted vthread's carrier support to ThreadsListHandle
- renamed fields/variables in GetStackTraceHandshakeClosure
- removed obsolete comment about executeOnCarrierThread

dholmes-ora commented:
> Added extra argument to cv_internal_thread_to_JavaThread to return carrrier's JavaThread.

@alexmenkov  I think we need a separate JBS issue for this and have runtime fix it. I want to be sure we are fixing it in the right/best way. I don't think we need the extra argument, but need to examine the existing usages. Chances are that any code trying to deal with virtual thread's via the API is actually doing it wrong and will need fixing anyway. I will file a bug.

dholmes-ora commented:
Filed [JDK-8361912](https://bugs.openjdk.org/browse/JDK-8361912)

alexmenkov commented:
> > Added extra argument to cv_internal_thread_to_JavaThread to return carrrier's JavaThread.
> 
> @alexmenkov I think we need a separate JBS issue for this and have runtime fix it. I want to be sure we are fixing it in the right/best way. I don't think we need the extra argument, but need to examine the existing usages. Chances are that any code trying to deal with virtual thread's via the API is actually doing it wrong and will need fixing anyway. I will file a bug.

I think just update the method to use carrier is error prone and callers should explicitly request the functionality.
I'm fine with separate issue, will remove changes in threadSMR.* and restore original handling of virtual thread carrier

dholmes-ora commented:
@alexmenkov  Alex, the form of this fix will be determined by whatever we end up doing for [JDK-8361912](https://bugs.openjdk.org/browse/JDK-8361912). I don't see any point in re-doing this fix later, so I suggest just putting this on hold for now. I will ensure the other issue is fixed within the next week.

dcubed-ojdk commented on src/hotspot/share/runtime/threadSMR.cpp:
```diff
@@ -791,14 +791,19 @@ ThreadsListHandle::~ThreadsListHandle() {
 // associated ThreadsList. This ThreadsListHandle "protects" the
 // returned JavaThread *.
 //
+// If the thread reference is a mounted virtual thread and use_carrier is true,
+// return JavaThread for the carrier thread; if the thread is unmounted
+// or use_carrier is false, return false.
+//
```
Right after the check on L827 below, what prevents the vthread from being unmounted?

So we fetch the JavaThread* on L833 and then verify that it is protected by the TLH
which is good, but that does not prevent the vthread from being unmounted by the
time that we return the JavaThread* on L853 below.

Am I missing something?

alexmenkov replied:
Yes, that's right.
vthread can be unmounted at any moment. So it's caller responsibility to check is vthread is still mounted to the same carrier thread in handshake/VMOp (see comment about virtul threads in the threadSMR.hpp)
`VirtualThread.asyncGetStackTrace` retries to get stack trace in the case

dholmes-ora replied:
@dcubed-ojdk No you are not missing anything, the vthread could become unmounted. It is up to the subsequent handshake code to re-check that (as briefly discussed above).

dholmes-ora replied:
Except I'm now discovering that can't work other than to abandon the virtual thread in question if you found the carrier has changed, otherwise you are in a handshake with the wrong carrier and so can't grab the stack - but that means you need to retry for this virtual thread.

dholmes-ora replied:
I see the current code just produces an empty stack trace if the virtual thread is found to have switched carriers. This seems wrong. We should be "pinning" the virtual thread whilst we try to get its stacktrace, such that it is unmounted or else mounted and fixed to a carrier. Other thread walking code uses the JvmtiVTMSTransitionDisabler to try and achieve this.

sspitsyn replied:
Sorry for being late to the party. I still need to walk through all the comments but it seems you have already caught the main problem with virtual threads. Nothing protects them from being concurrently mounted/unmounted/remounted. It is why the `JvmtiVTMSTransitionDisabler` is used in the JVMTI code. But, please, note that the `JvmtiVTMSTransitionDisabler` mechanism is enabled only when there is a JVMTI agent. Otherwise, it has been disabled for scalability purposes to exclude potentially high performance overhead at the VTMS transition points.

dholmes-ora replied:
If transition disabling is only available with JVMTI then I think we have a significant problem trying to deal with virtual threads in a number of APIs. Without disabling transitions you would basically need a loop that safely extracts the current carrier, handshakes with it, checks the VT is still mounted, and if not repeat the process. At the moment it appears that if the carrier is seen to have changed then we can just skip the VT - which would be very wrong IMO for most APIs e.g. thread dumps.

dholmes-ora replied:
> please, note that the JvmtiVTMSTransitionDisabler mechanism is enabled only when there is a JVMTI agent.

@sspitsyn  can you clarify this please. When you take a thread dump via the HotsotDiagnosticMXBean is there a JVMTI agent involved? Because that code uses the transition disabler.

AlanBateman replied:
Thread::getStackTrace handles this with a SUSPEND bit and loop. I think the only non-JVMTI API where we've got an issue right now is HotsotDiagnosticMXBean.dumpThreads.

sspitsyn replied:
> @sspitsyn can you clarify this please. When you take a thread dump via the HotsotDiagnosticMXBean is there a JVMTI agent involved? Because that code uses the transition disabler.

Yes, we have this issue for `HotsotDiagnosticMXBean.dumpThreads` as Alan has pointed out. We are aware about it and will evaluate a couple of fixing approaches.

dholmes-ora replied:
> Thread::getStackTrace handles this with a SUSPEND bit and loop.

Okay so that loop keeps retrying any time we get a null stacktrace from the VM, and unless we are extremely unlucky it should only take a couple of attempts to get an actual stack.

So that just leaves the issue of the carrier thread terminating, which is what this current fix was trying to address. So we need to ensure the carrier is protected by the TLH, which is what my other fix will address.

dholmes-ora replied:
BTW we need some comments making it clear that all null returns are handled via a retry at the higher level Java code.

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1882,33 +1882,26 @@ oop java_lang_Thread::park_blocker(oop java_thread) {
-    thread = java_lang_Thread::thread(java_thread);
-  }
-  if (thread == nullptr) {
+oop java_lang_Thread::async_get_stack_trace(jobject jthread, TRAPS) {
```
Please add a comment describing how the method works for virtual threads, and what returning null implies.

alexmenkov replied:
Done.

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1882,33 +1882,26 @@ oop java_lang_Thread::park_blocker(oop java_thread) {
   class GetStackTraceHandshakeClosure : public HandshakeClosure {
   public:
-    const Handle _java_thread;
+    const Handle _thread_oop;
```
Thanks for making this change! Very confusing otherwise.

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1930,21 +1923,22 @@ oop java_lang_Thread::async_get_stack_trace(oop java_thread, TRAPS) {
 
       bool carrier = false;
-      if (java_lang_VirtualThread::is_instance(_java_thread())) {
-        // if (thread->vthread() != _java_thread()) // We might be inside a System.executeOnCarrierThread
```
Need to check with @AlanBateman whether this check is also needed (he does both in his draft PR).

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1930,21 +1923,22 @@ oop java_lang_Thread::async_get_stack_trace(oop java_thread, TRAPS) {
+      if (java_lang_VirtualThread::is_instance(_thread_oop())) {
+        // Ensure _thread_oop is still mounted to java_thread.
+        const ContinuationEntry* ce = java_thread->vthread_continuation();
+        if (ce == nullptr || ce->cont_oop(java_thread) != java_lang_VirtualThread::continuation(_thread_oop())) {
+          // Target thread has been unmounted.
+          return;
         }
```
Shouldn't we set `carrier = true` if we don't return here? Or am I misunderstanding what "carrier" means here?

alexmenkov replied:
"carrier" means target is a platform thread with mounted virtual thread.
When `carrier == true` `vframeStream` skips frames of the mounted virtual thread.
Here target is a virtual thread, so we need frames starting from the last frame

sspitsyn reviewed (approved):
This looks good to me. I've posted one nit though.
